cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +224,153 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
   return left_shift<Rep, Rad>(val, scale);
 }
 
+/** @brief Extracts the sign, exponent, and integral significand of a floating point number.
+ *
+ * @note This returns all zero's for +/-0 and denormals, and zero's with an exponent of
+ * INT_MIN for the unrepresentable values +/-inf and NaN's.
+ *
+ * @tparam FloatingType Type of input floating point value
+ * @param floating The floating value to extract the components from
+ * @return A tuple containing the sign (bool), exponent (power-of-2), and integral significand.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline auto extract_components(FloatingType floating)
```

Comment:
Using frexp() and ilogb() are out of the question for doubles, as the entire conversion is 2x slower because they use the FP64 pipeline and the threads bottleneck on it.  Calling frexp() returns the mantissa as a floating point, and we need it as an integer during the conversion, so it's not really helping us, and so is slower even for floats.  I can try using ilogb() for floats just to see if the benchmarks are affected.

Reply:
Using ilogb() for floats is no faster, and increases code compexity as we'd now have separate code for floats and doubles.